circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+
+public class HiveMetaStoreClientCompatibility12x {
+
+  private final TServiceClient tServiceClient;
+
+  public HiveMetaStoreClientCompatibility12x(IMetaStoreClient client) {
+    while (Proxy.isProxyClass(client.getClass())) {
+      InvocationHandler handler = Proxy.getInvocationHandler(client);
+      if (handler.getClass().isAssignableFrom(RetryingMetaStoreClient.class)) {
+        client = getField(handler, "base");
+        continue;
+      }
+      // TODO add CloseableMetaStoreClientInvocationHandler ?
+      throw new RuntimeException("Unknown InvocationHandler " + handler.getClass());
+    }
+    tServiceClient = getField(client, "client");
+  }
+
+  private static <T> T getField(Object object, String fieldName) {
+    try {
+      Field field = object.getClass().getDeclaredField(fieldName);
+      T result = null;
+      if (field.isAccessible()) {
+        result = (T) field.get(object);
+      } else {
+        field.setAccessible(true);
+        result = (T) field.get(object);
+        field.setAccessible(false);
+      }
+      return result;
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to hack client", e);
+    }
+  }
+
+  private static Table deepCopy(Table table) {
+    return table.deepCopy();
+  }
+
+  private void sendBase(String methodName, TBase<?, ?> args) throws TException {
+    try {
+      Method sendBase = TServiceClient.class.getDeclaredMethod("sendBase", String.class, TBase.class);
+      sendBase.setAccessible(true);
+      sendBase.invoke(tServiceClient, methodName, args);
+      sendBase.setAccessible(false);
+    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException e) {
+      throw new RuntimeException("Unable to hack sendBase", e);
+    }
+  }
+
+  private void receiveBase(TBase<?, ?> result, String methodName) throws TException {
+    try {
+      Method receiveBase = TServiceClient.class.getDeclaredMethod("receiveBase", TBase.class, String.class);
+      receiveBase.setAccessible(true);
+      receiveBase.invoke(tServiceClient, result, methodName);
+      receiveBase.setAccessible(false);
+    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException e) {
+      throw new RuntimeException("Unable to hack receiveBase", e);
+    }
+  }
+
+  public Table getTable(String dbname, String name) throws MetaException, TException, NoSuchObjectException {
+    return deepCopy(get_table(dbname, name));
+  }
+
+  private Table get_table(String dbname, String tbl_name) throws MetaException, NoSuchObjectException, TException {
+    send_get_table(dbname, tbl_name);
+    return recv_get_table();
+  }
+
+  private void send_get_table(String dbname, String tbl_name) throws TException {
```

Comment:
Or is this and the method below the ones that have been copied from Hive?

Reply:
these were copied from Hive, I'll add a header on each method